crates/adapters/src/integrated/delta_table/input.rs

```diff
@@ -0,0 +1,688 @@
+                )
+                .await;
+            })
+        });
+
+        init_status_receiver.blocking_recv().ok_or_else(|| {
+            ControllerError::input_transport_error(
+                &endpoint.endpoint_name,
+                true,
+                anyhow!("worker thread terminated unexpectedly during initialization"),
+            )
+        })??;
+
+        Ok(Self { sender })
+    }
+
+    async fn worker_task(
+        endpoint: Arc<DeltaTableInputEndpointInner>,
+        input_stream: Box<dyn ArrowStream>,
+        receiver: Receiver<PipelineState>,
+        init_status_sender: mpsc::Sender<Result<(), ControllerError>>,
+    ) {
+        let mut receiver_clone = receiver.clone();
+        select! {
+            _ = Self::worker_task_inner(endpoint.clone(), input_stream, receiver, init_status_sender) => {
+                debug!("delta_table {}: worker task terminated",
+                    &endpoint.endpoint_name,
+                );
+            }
+            _ = receiver_clone.wait_for(|state| state == &PipelineState::Terminated) => {
+                debug!("delta_table {}: received termination command; worker task canceled",
+                    &endpoint.endpoint_name,
+                );
+            }
+        }
+    }
+
+    async fn worker_task_inner(
+        endpoint: Arc<DeltaTableInputEndpointInner>,
+        mut input_stream: Box<dyn ArrowStream>,
+        mut receiver: Receiver<PipelineState>,
+        init_status_sender: mpsc::Sender<Result<(), ControllerError>>,
+    ) {
+        let table = match endpoint.open_table().await {
+            Err(e) => {
+                let _ = init_status_sender.send(Err(e)).await;
+                return;
+            }
+            Ok(table) => table,
+        };
+
+        let table = Arc::new(table);
+
+        let snapshot_df = match endpoint.prepare_snapshot_query(&table).await {
+            Err(e) => {
+                let _ = init_status_sender.send(Err(e)).await;
+                return;
+            }
+            Ok(snapshot_df) => snapshot_df,
+        };
+
+        // Code before this point is part of endpoint initialization.
+        // After this point, the thread should continue running until it receives a
+        // shutdown command from the controller.
+        let _ = init_status_sender.send(Ok(())).await;
+
+        // Execute the snapshot query; push snapshot data to the circuit.
+        if let Some(snapshot_df) = snapshot_df {
+            endpoint
+                .execute_df(
+                    snapshot_df,
+                    true,
+                    "initial snapshot",
+                    input_stream.as_mut(),
+                    &mut receiver,
+                )
+                .await;
+
+            //let _ = endpoint.datafusion.deregister_table("snapshot");
+            info!(
+                "delta_table {}: finished reading initial snapshot",
+                &endpoint.endpoint_name,
+            );
```

Comment:
I think that reading a snapshot could take a long time if it's a big one, so I'd consider adding a log message just before starting that. It could be a good hint about where things are stuck when troubleshooting.